homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,126 @@
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve current app for %s", self.device_id)
-            return None
-
-    @property
-    def running_apps(self):
-        """Return a list of running apps."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'running'
-                    ), timeout=10).json()
-            return response.get('running_apps')
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve running apps for %s", self.device_id)
-            return None
-
-    def action(self, action_id):
-        """Perform an action on the device."""
-        try:
-            requests.get(DEVICE_ACTION_URL.format(
-                self.proto, self.host, self.port, self.device_id, action_id
-                ), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Action request for %s was not accepted for device %s",
-                action_id, self.device_id)
-
-    def start_app(self, app_name):
-        """Start an app."""
-        try:
-            requests.get(DEVICE_APPS_URL.format(
-                self.proto, self.host, self.port, self.device_id,
-                app_name + '/start'), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not start %s on %s", app_name, self.device_id)
+    adbkey = config.get(CONF_ADBKEY)
+    get_source = config.get(CONF_GET_SOURCE)
+    get_sources = config.get(CONF_GET_SOURCES)
+
+    device = FireTVDevice(host, name, adbkey, get_source, get_sources)
+    adb_log = " using adbkey='{0}'".format(adbkey) if adbkey else ""
+    if not device._firetv._adb:
+        _LOGGER.warning("Could not connect to Fire TV at %s%s", host, adb_log)
+
+        # Debugging
+        if adbkey != "":
+            # Check whether the key files exist
+            if not os.path.exists(adbkey):
+                raise FileNotFoundError(
+                    "ADB private key {} does not exist".format(adbkey))
+            if not os.path.exists(adbkey + ".pub"):
+                raise FileNotFoundError(
+                    "ADB public key {} does not exist".format(adbkey + '.pub'))
+
+            # Check whether the key files can be read
+            with open(adbkey):
+                pass
+            with open(adbkey + '.pub'):
+                pass
+
+    else:
+        _LOGGER.info("Setup Fire TV at %s%s", host, adb_log)
+        add_devices([device])
+
+
+def adb_wrapper(func):
+    """A wrapper that will wait if previous ADB commands haven't finished."""
+    @functools.wraps(func)
+    def _adb_wrapper(self, *args, **kwargs):
+        attempts = 0
+        while self._adb_lock and attempts < 5:
+            attempts += 1
+            time.sleep(1)
```

Comment:
Can you point me to an example or documentation for how to do that?

Reply:
The wemo switch platform does something similar.
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/wemo.py#L217-L234

A couroutine is needed to be able to await `asyncio.sleep`.

Here's our asyncio docs:
https://developers.home-assistant.io/docs/en/asyncio_index.html

Reply:
Thanks for those links. I think there's a more elegant solution than what I came up with... 

**The problem:** if you send a lot of commands in short succession, the ADB connection to the Fire TV gets broken. (This is a limitation of the `adb` Python package.)

**My solution:** use an `_adb_lock` attribute to indicate whether an ADB command is currently running, and if so, wait a second and then retry, with a max of 5 retries. 

I think this could be handled better using async, but I'm not sure how to do it.

Reply:
I don't know how to make async work, so I just removed the "wait 1 second and then retry" logic.